### DISCUSSION TITLE
If the ClientHello contains signature_algorithms_cert but does not contain signature_algorithms, the TLS1.2 handshake fails.

### DISCUSSION THREAD

**Discussion by lan1120** [openssl/openssl#22727] [Q&A] [answered]
**OpenSSL version：** 3.X

If the ClientHello contains signature_algorithms_cert but does not contain signature_algorithms, the TLS1.2 handshake fails.
**The server reports an error:** no shared signature algorithms.

**In the tls1_set_server_sigalgs function:**
If both peer_cert_sigalgs and peer_sigalgs are NULL, then searches for the default signature algorithm.
Otherwise, call tls1_process_sigalgs to obtain the shared signature algorithm, but the tls1_process_sigalgs function does not use signature_algorithms_cert. Therefore, the shared signature algorithm is always  NULL.

**Is this result expected?**

**Comment by davidben** [comment]
When signature_algorithms is omitted, TLS 1.2 (foolishly...) defined the client as advertising only SHA-1. I believe OpenSSL disables SHA-1 by default now, so that may be the cause.

**Comment by mattcaswell** [comment, marked as answer]
> If the ClientHello contains signature_algorithms_cert but does not contain signature_algorithms, the TLS1.2 handshake fails.
The server reports an error: no shared signature algorithms.

signature_algorithms are optional in TLSv1.2. If they are not present then OpenSSL should fallback to "default" signature algorithms. This is specified in RFC5246:

````
   If the client does not send the signature_algorithms extension, the
   server MUST do the following:

   -  If the negotiated key exchange algorithm is one of (RSA, DHE_RSA,
      DH_RSA, RSA_PSK, ECDH_RSA, ECDHE_RSA), behave as if client had
      sent the value {sha1,rsa}.

   -  If the negotiated key exchange algorithm is one of (DHE_DSS,
      DH_DSS), behave as if the client had sent the value {sha1,dsa}.

   -  If the negotiated key exchange algorithm is one of (ECDH_ECDSA,
      ECDHE_ECDSA), behave as if the client had sent value {sha1,ecdsa}.
````

All of the default sig algs are based on SHA1 - which OpenSSL disallows in its default security level. In that case I would expect the "no shared signature algorithms" message that you are seeing. If the default security level is set to 0 (e.g. by setting "@SECLEVEL=0" at the end of the ciphersuites list), then a successful handshake should still be possible.

"signature_algorithms_cert" only specifies the sig algs to be used in signing certs. It is never used for handshake signatures.

> Is this result expected?

Yes

**Reply by lan1120** [reply]
I've set the security level to 0 during the test.
In tls1_set_server_sigalgs function, tls1_process_sigalgs failed, not tls1_get_legacy_sigalg.

**Reply by lan1120** [reply]
@mattcaswell The function tls1_set_server_sigalgs searches for the default signature algorithm only when both peer_cert_sigalgs and peer_sigalgs are NULL.
